### ISSUE TITLE
[FR] Add support for RISC-V open ISA CPUs

### ISSUE THREAD

**Issue by mark-stopka** [xmrig/xmrig#1924]
Hi, RISC-V is gaining traction, and more and more powerfull open IP CPUs are becomming available on the market, I was wondering if you guys considered adding RISC-V support for RandomX into `xmrig`.

**Comment by SChernykh** [comment]
You are probably talking about https://www.sifive.com/blog/the-heart-of-risc-v-development-is-unmatched - yes, it will be able to run RandomX with decent hashrate, but it's a bit expensive at the moment. RISC-V is in the wish list, but there are no solid plans for it now. If you know of any RISC-V board with 4+ GB RAM and 64-bit CPU at no more than $100, let us know.

**Comment by mark-stopka** [comment]
I may be willing to sponsor one for development purposes, are you US based so there would be no additional import dutties?

**Comment by SChernykh** [comment]
I'm in Europe. I'll be busy with Zen3-related changes in November and probably December anyway. Maybe RISC-V availability will be better in 2021.

**Comment by mark-stopka** [comment]
Ok, I'll ping back in 2021, if you could leave the ticket open I would appreciate it, as I link it in my private Issue Tracker...

**Comment by Lonnegan** [comment]
> I'll be busy with Zen3-related changes in November and probably December anyway. 

Okay... Will there be changes in Zen 3 which will affect xmrig? AFAIK will the cache sizes stay the same (hence in different structure (1x 32 MB per die instead if 2x 16 MB per die), but no new instruction sets. Of have you other information?

**Comment by SChernykh** [comment]
Zen3 has vector AES instructions, I need to add support for them. MSR mod might not work as intended because it's CPU specific. Plus I need to check all previous Zen1/Zen2 optimizations to see if they work or not.

**Comment by Lonnegan** [comment]
@SChernykh Did you have the chance to test the impact of VAES to one of the mining algo, already?

**Comment by SChernykh** [comment]
Yes, I tested 2 different ways to use VAES and they both didn't give any speedup. AES in RandomX is limited by instruction latency and it stayed the same for VAES instructions.

**Comment by Lonnegan** [comment]
> AES in RandomX is limited by instruction latency and it stayed the same for VAES instructions.

Plus the higher latency of the L3 cache in comparison to Zen 2. Seems like Zen 3 is not the same game changer in RandomX mining as it is in gaming then.

**Comment by xmrig** [comment]
@mark-stopka Any news here? Thank you.

**Comment by mark-stopka** [comment]
> @mark-stopka Any news here? Thank you.

I am not sure what news should I have, there is still only one board available, and I already offered to sponsor one for you guys if you'd like to implement RISC-V support, however, there should be new boards and chips available in next 6 months, so:

a) you'd like me to sponsor the current SiFive board which I am happy to do
b) wait for new boards and re-evaluate
c) you expect me to hire a 3rd party to implement such support, in which case I can look for someone, but having it done by the core team would probably be more cost-effective as you already know the codebase and the caveats of RandomX mining algo, I would have to recruit someone who never implemented RandomX before, nor contributed to your codebase as don't know anyone who has that experience

Let me know which one works for you, and we can move forward (or wait in case of option 2)

**Comment by SChernykh** [comment]
I did a quick look again at https://www.sifive.com/boards/hifive-unmatched and https://www.sifive.com/cores/u74-mc and it seems that this board will be able to run 1 RandomX thread (2 MB L2 cache) at the full speed with JIT enabled (RV64GBC support covers all RandomX instructions), so it's good enough for development. Also this quote
```
The L1 Instruction Cache and the L2 Cache can be configured into high-speed deterministic SRAMs
```
got me really curious on how it can be used for RandomX. Price of the board is still quite high, but if you're happy to send it my way to Sweden (I can pay import tax), I'll add full support to xmrig.

**Comment by mark-stopka** [comment]
@SChernykh, cool, can you ping me at mstopka@opensuse.org?

**Comment by gavan1** [comment]
@mark-stopka @SChernykh I was one of the few that received a BeagleV board here in Tokyo. Using the [July 27th fedora release](https://github.com/starfive-tech/Fedora_on_StarFive) 

Following the Fedora build steps, dependencies installed correctly, 

`make` error.

```
[riscv@fedora-starfive build]$ make -j2
Scanning dependencies of target ethash
Scanning dependencies of target xmrig-asm
[  0%] Building C object src/3rdparty/libethash/CMakeFiles/ethash.dir/ethash_internal.c.o
[  1%] Building ASM object CMakeFiles/xmrig-asm.dir/src/crypto/cn/asm/cn_main_loop.S.o
cc: error: unrecognized command-line option ‘-maes’
make[2]: *** [src/3rdparty/libethash/CMakeFiles/ethash.dir/build.make:82: src/3rdparty/libethash/CMakeFiles/ethash.dir/ethash_internal.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:375: src/3rdparty/libethash/CMakeFiles/ethash.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  1%] Building ASM object CMakeFiles/xmrig-asm.dir/src/crypto/cn/asm/CryptonightR_template.S.o
/home/riscv/xmrig-6.12.2-mo2/src/crypto/cn/asm/cn_main_loop.S: Assembler messages:
/home/riscv/xmrig-6.12.2-mo2/src/crypto/cn/asm/cn_main_loop.S:6: Error: unknown pseudo-op: `.intel_syntax'
/home/riscv/xmrig-6.12.2-mo2/src/crypto/cn/asm/cn_main_loop.S:22: Warning: alignment too large: 63 assumed

as: out of memory allocating 9223372036854841442 bytes after a total of 135168 bytes
make[2]: *** [CMakeFiles/xmrig-asm.dir/build.make:82: CMakeFiles/xmrig-asm.dir/src/crypto/cn/asm/cn_main_loop.S.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/riscv/xmrig-6.12.2-mo2/src/crypto/cn/asm/CryptonightR_template.S: Assembler messages:
/home/riscv/xmrig-6.12.2-mo2/src/crypto/cn/asm/CryptonightR_template.S:6: Error: unknown pseudo-op: `.intel_syntax'
/home/riscv/xmrig-6.12.2-mo2/src/crypto/cn/asm/CryptonightR_template.inc:13: Warning: alignment too large: 63 assumed

as: out of memory allocating 9223372036854841442 bytes after a total of 270336 bytes
make[2]: *** [CMakeFiles/xmrig-asm.dir/build.make:87: CMakeFiles/xmrig-asm.dir/src/crypto/cn/asm/CryptonightR_template.S.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:182: CMakeFiles/xmrig-asm.dir/all] Error 2
make: *** [Makefile:103: all] Error 2
```

**Comment by benthetechguy** [comment]
I'm also interested in RISC-V xmrig, I have a Nezha D1 I'm willing to test any code on.

**Comment by divinity76** [comment]
@gavan1 and what happens if you compile with
```
-DWITH_CN_LITE=OFF -DWITH_CN_HEAVY=OFF -DWITH_CN_PICO=OFF -DWITH_CN_FEMTO=OFF
```
?

**Comment by kami4ka** [comment]
@divinity76 nothing, the code won't compile due to the lack of code for RISC-V, so it tries to build x86 instead

**Comment by KiritakeKumi** [comment]
Currently, there is a host milkv pioneer on the market that uses the high-performance RISC-V chip SG2042. This chip looks very similar to the Antminer X5 dedicated XMR mining machine. Has anyone done research on this device?

**Comment by SChernykh** [comment]
Antminer X5 uses SG2042R which is a version of SG2042 with some additional instructions and hardware AES. I guess they added new instructions for the most commonly used sequences in RandomX generated code.

**Comment by KiritakeKumi** [comment]
So can we set an option in the compilation to support RISC-V without any acceleration to avoid compilation failures due to x86/arm related instruction sets, later add RISC-V Cryptography Extension (K) to achieve full acceleration support.

**Comment by SChernykh** [comment]
It's not supported by XMRig yet. Once I have an easily accessible RISC-V system, I will make sure it compiles.

**Comment by KiritakeKumi** [comment]
RISC-V International provides some RISC-V devices that can be accessed remotely. If you don’t have any devices, you can choose here https://riscv.org/risc-v-labs/.

**Comment by aragubas** [comment]
I have a StarFive VisionFive 2 with 8GB of RAM and Dual Ethernet model, any updates on this?

**Comment by aragubas** [comment]
Trying to build on the StarFive VisionFive 2 yields the following errors:

```
[  2%] Building C object src/3rdparty/argon2/CMakeFiles/argon2-avx2.dir/arch/x86_64/lib/argon2-avx2.c.o
[  2%] Building C object src/3rdparty/libethash/CMakeFiles/ethash.dir/ethash_internal.c.o
[  2%] Building ASM object CMakeFiles/xmrig-asm.dir/src/crypto/cn/asm/cn_main_loop.S.o
[  2%] Building C object src/crypto/ghostrider/CMakeFiles/ghostrider.dir/sph_blake.c.o
cc: error: unrecognized command-line option ‘-maes’
cc: error: unrecognized command-line option ‘-maes’
make[2]: *** [src/3rdparty/argon2/CMakeFiles/argon2-avx2.dir/build.make:76: src/3rdparty/argon2/CMakeFiles/argon2-avx2.dir/arch/x86_64/lib/argon2-avx2.c.o] Error 1
make[2]: *** [src/3rdparty/libethash/CMakeFiles/ethash.dir/build.make:76: src/3rdparty/libethash/CMakeFiles/ethash.dir/ethash_internal.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:368: src/3rdparty/libethash/CMakeFiles/ethash.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:286: src/3rdparty/argon2/CMakeFiles/argon2-avx2.dir/all] Error 2
cc: error: unrecognized command-line option ‘-maes’
make[2]: *** [src/crypto/ghostrider/CMakeFiles/ghostrider.dir/build.make:76: src/crypto/ghostrider/CMakeFiles/ghostrider.dir/sph_blake.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/dietpi/Downloads/xmrig/src/crypto/cn/asm/cn_main_loop.S: Assembler messages:
/home/dietpi/Downloads/xmrig/src/crypto/cn/asm/cn_main_loop.S:6: Error: unknown pseudo-op: `.intel_syntax'
/home/dietpi/Downloads/xmrig/src/crypto/cn/asm/cn_main_loop.S:25: Warning: alignment too large: 63 assumed

as: out of memory allocating 9223372036854841462 bytes after a total of 270336 bytes
make[2]: *** [CMakeFiles/xmrig-asm.dir/build.make:75: CMakeFiles/xmrig-asm.dir/src/crypto/cn/asm/cn_main_loop.S.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[  3%] Building ASM object CMakeFiles/xmrig-asm.dir/src/crypto/cn/asm/CryptonightR_template.S.o
[  3%] Building C object src/crypto/ghostrider/CMakeFiles/ghostrider.dir/sph_bmw.c.o
cc: error: unrecognized command-line option ‘-maes’
make[2]: *** [src/crypto/ghostrider/CMakeFiles/ghostrider.dir/build.make:90: src/crypto/ghostrider/CMakeFiles/ghostrider.dir/sph_bmw.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:394: src/crypto/ghostrider/CMakeFiles/ghostrider.dir/all] Error 2
/home/dietpi/Downloads/xmrig/src/crypto/cn/asm/CryptonightR_template.S: Assembler messages:
/home/dietpi/Downloads/xmrig/src/crypto/cn/asm/CryptonightR_template.S:6: Error: unknown pseudo-op: `.intel_syntax'
/home/dietpi/Downloads/xmrig/src/crypto/cn/asm/CryptonightR_template.inc:13: Warning: alignment too large: 63 assumed

as: out of memory allocating 9223372036854841462 bytes after a total of 405504 bytes
make[2]: *** [CMakeFiles/xmrig-asm.dir/build.make:88: CMakeFiles/xmrig-asm.dir/src/crypto/cn/asm/CryptonightR_template.S.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:148: CMakeFiles/xmrig-asm.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

**Comment by benthetechguy** [comment]
That error just means it's trying to compile with x86-specific options, because xmrig is built to treat everything it can't detect ARM on as x86.

**Comment by aragubas** [comment]
> That error just means it's trying to compile with x86-specific options, because xmrig is built to treat everything it can't detect ARM on as x86.

Is there a way for me to specify that I'm building on RISC-V?

**Comment by benthetechguy** [comment]
No, there's no code for RISC-V yet. There's code in xmrig that uses x86 specific features, and the only reason ARM works is because of translation layers like sse2neon. Before you can compile for RISC-V, we need to implement something like that.

**Comment by SChernykh** [comment]
Not yet, but I think it can be fixed with a few changes in CMake files. I'll try to set up a risc-v Qemu VM to see what needs to be done.

**Comment by SChernykh** [comment]
https://github.com/tevador/RandomX/ already supports RISC-V, and https://github.com/SChernykh/p2pool/ uses it so it also supports RISC-V.

XMRig just has many many changes which will make integration of that code not straight-forward.

**Comment by IngwiePhoenix** [comment]
What exactly would be needed for said integration?

I often find myself experimenting with RISC-V and am in the process of aquiring a Milk-V Pioneer - which has a pretty good compute grunt. So I might be able to help out - but I am entirely new to the project and it's structure...though I do bring experience in C and Go.
